cumulus/pallets/collator-selection/src/lib.rs

```diff
@@ -554,6 +630,156 @@ pub mod pallet {
+		}
+
+		/// The caller `origin` replaces a candidate `target` in the collator candidate list by
+		/// reserving `deposit`. The amount `deposit` reserved by the caller must be greater than
+		/// the existing bond of the target it is trying to replace.
+		///
+		/// This call will fail if the caller is already a collator candidate or invulnerable, the
+		/// caller does not have registered session keys, the target is not a collator candidate,
+		/// and/or the `deposit` amount cannot be reserved.
+		#[pallet::call_index(8)]
+		#[pallet::weight(T::WeightInfo::take_candidate_slot(T::MaxCandidates::get()))]
+		pub fn take_candidate_slot(
+			origin: OriginFor<T>,
+			deposit: BalanceOf<T>,
+			target: T::AccountId,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			ensure!(!Self::invulnerables().contains(&who), Error::<T>::AlreadyInvulnerable);
+			ensure!(deposit >= Self::candidacy_bond(), Error::<T>::InsufficientBond);
+
+			let validator_key = T::ValidatorIdOf::convert(who.clone())
+				.ok_or(Error::<T>::NoAssociatedValidatorId)?;
+			ensure!(
+				T::ValidatorRegistration::is_registered(&validator_key),
+				Error::<T>::ValidatorNotRegistered
+			);
+
+			ensure!(deposit >= Self::candidacy_bond(), Error::<T>::InsufficientBond);
+
+			let length = <CandidateList<T>>::decode_len().unwrap_or_default();
+			// The closure below iterates through all elements of the candidate list to ensure that
+			// the caller isn't already a candidate and to find the target it's trying to replace in
+			// the list. The return value is a tuple of the position of the candidate to be replaced
+			// in the list along with its candidate information.
+			let target_info = <CandidateList<T>>::try_mutate(
+				|candidates| -> Result<CandidateInfo<T::AccountId, BalanceOf<T>>, DispatchError> {
+					// Find the position in the list of the candidate that is being replaced.
+					let mut target_info_idx = None;
+					let mut new_info_idx = None;
+					for (idx, candidate_info) in candidates.iter().enumerate() {
+						// While iterating through the candidates trying to find the target,
+						// also ensure on the same pass that our caller isn't already a
+						// candidate.
+						ensure!(candidate_info.who != who, Error::<T>::AlreadyCandidate);
+						// If we find our target, update the position but do not stop the
+						// iteration since we're also checking that the caller isn't already a
+						// candidate.
+						if candidate_info.who == target {
+							target_info_idx = Some(idx);
+						}
+						// Find the spot where the new candidate would be inserted in the current
+						// version of the list.
+						if new_info_idx.is_none() && candidate_info.deposit >= deposit {
+							new_info_idx = Some(idx);
+						}
+					}
+					let target_info_idx =
+						target_info_idx.ok_or(Error::<T>::TargetIsNotCandidate)?;
+
+					// Remove the old candidate from the list.
+					let target_info = candidates.remove(target_info_idx);
```

Comment:
why removing? should not it just move one position below? `target` might still have enough to be elected, and why it should register itself again?

Reply:
this might serve the case when there is no slot left for a new candidate. but the might have candidates with lower deposits.

Reply:
This is intended behavior for this extrinsic. Anyone can kick out anyone as long as they can put up more funds than the targeted candidate. The economic incentive for candidates is to bid as low as possible to get in anyway, so it doesn't make much of a difference.

Reply:
yes, this works. but it is intended like that? I wonder why we would let kick someone till the list is not full, and why not from the bottom.

Reply:
> I wonder why we would let kick someone till the list is not full

We could disallow it, but it's not in the economic best interest of a user to use `take_candidate_slot` when they could `register_as_candidate` with the minimum bond.

> why not from the bottom

This is a consequence of having the API and behavior of the extrinsic compatible with the different approaches discussed in https://github.com/paritytech/cumulus/pull/2960, namely keeping a lazily sorted list. It's obvious it would be more difficult to maintain the same behavior if the list is lazily sorted and would defeat the purpose. It's not *necessary* to keep the same API of behavior right now, but we made the decision to do so because it's not expensive, allows for easy change in the future and would keep the same behavior across implementations. Also, this implementation favors the correct behavior anyway through the economic incentives, so we reach the same end result but we keep our options open for the future.